src/dsp_tools/utils/xmlupload/xmlupload.py

```diff
@@ -345,6 +417,153 @@ def _upload_resources(
     return id2iri_mapping, failed_uploads, metrics
 
 
+def _upload_multimedia_to_sipi(
+    resource: XMLResource,
+    sipi_server: Sipi,
+    imgdir: str,
+    filesize: float,
+    permissions_lookup: dict[str, Permissions],
+    metrics: list[MetricRecord],
+    preprocessing_done: bool,
+) -> dict[str, str | Permissions] | None:
+    """
+    This function takes a resource with a corresponding bitstream filepath.
+    If the pre-processing is not done, it retrieves the file from the directory and uploads it to sipi.
+    If pre-processing is done it retrieves the bitstream information from sipi.
+
+    Args:
+        resource: resource with that has a bitstream
+        sipi_server: server to upload
+        imgdir: directory of the file
+        filesize: size of the file
+        permissions_lookup: dictionary that contains the permission name as string and the corresponding Python object
+        metrics: to store metric information in
+        preprocessing_done: If True, then no upload is necessary
+
+    Returns:
+        The information from sipi which is needed to establish a link from the resource
+    """
+    if preprocessing_done:
+        resource_bitstream = resource.get_bitstream(
+            internal_file_name_bitstream=resource.bitstream.value,  # type: ignore[union-attr]
+            permissions_lookup=permissions_lookup,
+        )
```

Comment:
I would factor that out to a function with a nice name (`_get_bitstream_information_from_sipi()`)